packages/automerge-repo/src/Presence.ts

```diff
@@ -0,0 +1,592 @@
+   * @returns Array of peer ids for this user
+   */
+  getUserPeers(userId: UserId) {
+    const peers = this.#userPeers.get(userId)
+    if (!peers) {
+      return
+    }
+    return Array.from(peers)
+  }
+
+  /**
+   * Get all recently-seen peers for this device.
+   *
+   * @param deviceId
+   * @returns Array of peer ids for this device
+   */
+  getDevicePeers(deviceId: DeviceId) {
+    const peers = this.#devicePeers.get(deviceId)
+    if (!peers) {
+      return
+    }
+    return Array.from(peers)
+  }
+
+  /**
+   * Get most-recently-seen peer from this group.
+   *
+   * @param peers
+   * @returns id of most recently seen peer
+   */
+  getFreshestPeer(peers: Set<PeerId>) {
+    let freshestLastSeen: number
+    return Array.from(peers).reduce((freshest: PeerId | undefined, curr) => {
+      const lastSeen = this.#peersLastSeen.get(curr)
+      if (!lastSeen) {
+        return freshest
+      }
+
+      if (!freshest || lastSeen > freshestLastSeen) {
+        freshestLastSeen = lastSeen
+        return curr
+      }
+
+      return freshest
+    }, undefined)
+  }
+
+  /**
+   * Get current @type PeerState for given peer.
+   *
+   * @param peerId
+   * @returns details for the peer
+   */
+  getPeerInfo(peerId: PeerId) {
+    return this.#peerStates.get(peerId)
+  }
+
+  /**
+   * Get current ephemeral state value for this peer. If a channel is specified,
+   * only returns the ephemeral state for that specific channel. Otherwise,
+   * returns the full ephemeral state.
+   *
+   * @param peerId
+   * @param channel
+   * @returns latest ephemeral state received
+   */
+  getPeerState<Channel extends keyof State>(peerId: PeerId, channel?: Channel) {
+    const fullState = this.#peerStates.get(peerId)?.value
+    if (!channel) {
+      return fullState
+    }
+
+    return fullState?.[channel]
+  }
+
+  /**
+   * Get current ephemeral state value for this user's most-recently-active
+   * peer. See {@link getPeerState}.
+   *
+   * @param userId
+   * @param channel
+   * @returns
+   */
+  getUserState<Channel extends keyof State>(userId: UserId, channel?: Channel) {
+    const peers = this.#userPeers.get(userId)
+    if (!peers) {
+      return undefined
+    }
+    const peer = this.getFreshestPeer(peers)
+    if (!peer) {
+      return undefined
+    }
+
+    return this.getPeerState(peer, channel)
+  }
+
+  /**
+   * Get current ephemeral state value for this device's most-recently-active
+   * peer. See {@link getPeerState}.
+   *
+   * @param userId
+   * @param channel
+   * @returns
+   */
+  getDeviceState<Channel extends keyof State>(
+    deviceId: UserId,
+    channel?: Channel
+  ) {
+    const peers = this.#devicePeers.get(deviceId)
+    if (!peers) {
+      return undefined
+    }
+    const peer = this.getFreshestPeer(peers)
+    if (!peer) {
+      return undefined
+    }
+
+    return this.getPeerState(peer, channel)
+  }
+}
+
+class PeerPresenceInfo<State> extends EventEmitter<PresenceEvents> {
```

Comment:
This class is used internally by Presence to manage the state that's exposed to users through `PeerPresenceView` above.